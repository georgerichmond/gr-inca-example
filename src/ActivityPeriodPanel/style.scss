.inca-period-picker-panel {
  .ant-radio-group {
    font-size: 1.3rem;
    display: flex;
  }

  .ant-radio-button-wrapper {
    flex-grow: 1;
    text-align: center;
  }

  .ant-menu-horizontal {
    line-height: 3rem;
    display: flex;
  }

  .ant-menu-item {
    flex: 0 0 10%;
    text-align: center;
    padding: 0 8px;
    &.custom-date-picker-list {
      flex: 1 0 10%;
    }
    &.feature-toggle-off {
      flex-grow: 1;
    }
  }

  .custom-datepicker-list {
    padding: 0;
  }
}
.custom-date-picker {
  .ant-calendar-picker-input {
    border: 0;
    background: transparent;
    font-size: 14px;
    padding: 4px 0;
  }
  .ant-calendar-picker-clear {
    position: relative;
    right: -6px;
    top: 0;
  }
  .ant-calendar-picker-icon {
    position: relative;
    top: 0;
    right: 0;
  }
  .ant-calendar-range-picker-input {
    width: 40%;
  }
}

.custom-date-picker-drop {
  .ant-calendar-range {
    width: 100%;
    .ant-calendar-input {
      padding-left: 10px;
    }
  }

  .ant-calendar-range-right {
    .ant-calendar-date-input-wrap {
      margin-left: 0;
    }
  }

  .ant-calendar-range-middle {
    left: 47.5%;
    margin-left: 0;
  }

  .ant-calendar {
    border: 5px solid #19374c;
  }

  .ant-calendar-disabled-cell {
    opacity: 0.2;
  }
}
