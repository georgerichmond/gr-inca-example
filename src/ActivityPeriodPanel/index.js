// @flow
import React from 'react';
import { Menu, DatePicker } from 'antd';
import { getOr } from 'lodash/fp';
import moment from 'moment';
import { times } from '../../helpers/activityPeriod';
import apolloClient from '../../apollo/client';
import {
  getActivityPeriodGraphQLDataFromOption,
  disabledDate
} from '../../helpers/activityPeriod';
import './style.scss';
import ReactGA from 'react-ga';

export const INCA_GROUP = 'inca-group';

type AppMetaData = {
  role: string,
  features: Array<string>
};
type Props = {
  activityPeriodOption: string,
  metaData: AppMetaData
};
type State = {
  defaultActivityPeriod: Array<string>
};

type OnActivityPeriodChange = (string | Array<moment>) => void;

type setPickerValue = Object => void;

class ActivityPeriodPanel extends React.Component<Props, State> {
  state = {
    defaultActivityPeriod: [moment().startOf('day'), moment().endOf('day')]
  };

  onActivityPeriodChange: OnActivityPeriodChange = option => {
    const label = Array.isArray(option) ? 'custom' : option;
    ReactGA.event({
      category: 'Activity period',
      action: 'Change option',
      label
    });
    const data = getActivityPeriodGraphQLDataFromOption(option);
    this.setPickerValue({ data, option });
    apolloClient.writeData({
      data
    });
  };

  setPickerValue: setPickerValue = ({ data, option }) => {
    const { from, to } = data.activityPeriod;
    if (Array.isArray(option))
      this.setState({
        defaultActivityPeriod: [
          moment(from).startOf('day'),
          moment(to).endOf('day')
        ]
      });
    else
      this.setState({
        defaultActivityPeriod: [
          moment(from).startOf('day'),
          moment(to)
            .subtract(1, 'day')
            .endOf('day')
        ]
      });
  };
  render() {
    const { activityPeriodOption, metaData } = this.props;

    console.log(metaData)



    const isINCAGroupMember = getOr('', 'role', metaData).includes(INCA_GROUP);

    return (
      <div className="inca-period-picker-panel" data-test="period-panel">
        <Menu mode="horizontal" selectedKeys={[activityPeriodOption]}>
          {times.map(({ value, label }) => (
            <Menu.Item
              key={value}
              data-test={value}
              onClick={({ key }) => this.onActivityPeriodChange(key)}
              className={!isINCAGroupMember ? 'feature-toggle-off' : ''}
            >
              {label}
            </Menu.Item>
          ))}
          {isINCAGroupMember && (
            <Menu.Item
              key="custom"
              data-test="custom"
              className="custom-date-picker-list"
            >
              <DatePicker.RangePicker
                className="custom-date-picker"
                dropdownClassName="custom-date-picker-drop"
                id="period-selector"
                format="ddd D MMM"
                value={this.state.defaultActivityPeriod}
                onChange={this.onActivityPeriodChange}
                disabledDate={disabledDate}
                allowClear={false}
              />
            </Menu.Item>
          )}
        </Menu>
      </div>
    );
  }
}
export default ActivityPeriodPanel;
