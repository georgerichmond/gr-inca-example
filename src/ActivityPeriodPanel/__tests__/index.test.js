import React from 'react';
import { shallow } from 'enzyme';
import { Menu } from 'antd';
import ActivityPeriodPanel, { INCA_GROUP } from '../';
import apolloClient from '../../../apollo/client';
import { getActivityPeriodGraphQLDataFromOption } from '../../../helpers/activityPeriod';
import ReactGA from 'react-ga';

jest.mock('../../../helpers/activityPeriod', () => ({
  times: [{ value: 1, label: 'Today' }, { value: 2, label: 'Yesterday' }],
  getActivityPeriodGraphQLDataFromOption: jest.fn()
}));

describe('Activity Period Panel', () => {
  beforeAll(() => {
    getActivityPeriodGraphQLDataFromOption.mockImplementation(option => ({
      activityPeriod: {
        from: '2018-09-01T00:00:00Z',
        to: '2018-09-01T23:59:59Z'
      },
      option
    }));
  });

  it('renders correctly', () => {
    const result = shallow(
      <ActivityPeriodPanel
        activityPeriodOption="today"
        metaData={{ role: INCA_GROUP, features: [] }}
      />
    );

    expect(result).toMatchSnapshot();
  });

  it('calls the onActivityPeriodChange callback when an option is selected', () => {
    const wrapper = shallow(
      <ActivityPeriodPanel
        activityPeriodOption="today"
        metaData={{ role: INCA_GROUP, features: [] }}
      />
    );
    const activityPeriodChange = jest.spyOn(
      wrapper.instance(),
      'onActivityPeriodChange'
    );
    const onClickHandler = wrapper
      .find(Menu.Item)
      .first()
      .prop('onClick');

    onClickHandler({ key: 'today' });

    expect(activityPeriodChange).toHaveBeenCalledWith('today');
  });

  describe('Activity Period change', () => {
    beforeAll(() => {
      jest.spyOn(apolloClient, 'writeData');
      jest.spyOn(ReactGA, 'event').mockImplementation(() => {});
      getActivityPeriodGraphQLDataFromOption.mockImplementation(option => ({
        activityPeriod: {
          from: '2018-09-01T00:00:00Z',
          to: '2018-09-01T23:59:59Z'
        },
        option
      }));
      const wrapper = shallow(
        <ActivityPeriodPanel
          activityPeriodOption="today"
          metaData={{ role: INCA_GROUP, features: [] }}
        />
      );

      const onClickHandler = wrapper
        .find(Menu.Item)
        .first()
        .prop('onClick');

      onClickHandler({ key: 'today' });
    });

    it('writes the selected activity period and option to the Apollo client', () => {
      expect(apolloClient.writeData).toHaveBeenCalledWith({
        data: {
          activityPeriod: {
            from: '2018-09-01T00:00:00Z',
            to: '2018-09-01T23:59:59Z'
          },
          option: 'today'
        }
      });
    });

    it('tracks the event in google analytics', () => {
      expect(ReactGA.event).toHaveBeenCalledWith({
        action: 'Change option',
        category: 'Activity period',
        label: 'today'
      });
    });
  });
});
