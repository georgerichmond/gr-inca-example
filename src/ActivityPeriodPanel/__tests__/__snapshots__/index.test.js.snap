// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity Period Panel renders correctly 1`] = `
<div
  className="inca-period-picker-panel"
  data-test="period-panel"
>
  <Menu
    className=""
    focusable={false}
    mode="horizontal"
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        "today",
      ]
    }
    theme="light"
  >
    <MenuItem
      className=""
      data-test={1}
      key="1"
      onClick={[Function]}
    >
      Today
    </MenuItem>
    <MenuItem
      className=""
      data-test={2}
      key="2"
      onClick={[Function]}
    >
      Yesterday
    </MenuItem>
    <MenuItem
      className="custom-date-picker-list"
      data-test="custom"
      key="custom"
    >
      <PickerWrapper
        allowClear={false}
        className="custom-date-picker"
        dropdownClassName="custom-date-picker-drop"
        format="ddd D MMM"
        id="period-selector"
        inputPrefixCls="ant-input"
        locale={Object {}}
        onChange={[Function]}
        onOk={[Function]}
        onOpenChange={[Function]}
        popupStyle={Object {}}
        prefixCls="ant-calendar"
        transitionName="slide-up"
        value={
          Array [
            "2019-06-20T23:00:00.000Z",
            "2019-06-21T22:59:59.999Z",
          ]
        }
      />
    </MenuItem>
  </Menu>
</div>
`;
